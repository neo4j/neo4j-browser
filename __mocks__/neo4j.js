/*
 * Copyright (c) 2002-2018 "Neo4j, Inc"
 * Network Engine for Objects in Lund AB [http://neotechnology.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

function integerFn (val) {
  this.val = val
}
integerFn.prototype.toString = function () {
  return this.val.toString()
}

var out = {
  v1: {
    Int: integerFn,
    isInt: function (val) {
      return val instanceof integerFn
    },
    types: {
      Node: function Node (id, labels, properties) {
        this.identity = id
        this.labels = labels
        this.properties = properties
      },
      Relationship: function Relationship (id, start, end, type, properties) {
        this.identity = id
        this.start = start
        this.end = end
        this.type = type
        this.properties = properties
      },
      Path: function Path (start, end, segments) {
        this.start = start
        this.end = end
        this.segments = segments
        this.length = segments.length
      },
      PathSegment: function PathSegment (start, relationship, end) {
        this.start = start
        this.relationship = relationship
        this.end = end
      }
    },
    Integer: function Integer ({ low, high }) {
      this.low = low
      this.high = high
    }
  }
}

out.v1.types.Node.prototype.toString = function () {
  return 'node'
}
out.v1.types.Relationship.prototype.toString = function () {
  return 'rel'
}
out.v1.types.Path.prototype.toString = function () {
  return 'path'
}
out.v1.types.PathSegment.prototype.toString = function () {
  return 'pathsegment'
}
out.v1.Integer.prototype.toInt = function () {
  return this.low
}
out.v1.int = val => new out.v1.Integer(val)

module.exports = out
