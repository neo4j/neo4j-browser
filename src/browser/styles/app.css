/* Base imports */
@import './base/reset.css';
@import './base/variables.css';
@import './base/fonts.css';

/* Component styles */
@import './components/buttons.css';
@import './components/forms.css';
@import './components/icons.css';

/* Theme styles - loaded dynamically based on theme */
@import './themes/light.css' (prefers-color-scheme: light);
@import './themes/dark.css' (prefers-color-scheme: dark);

/* Essential utility classes that extend Tailwind */
.neo4j-mono {
  font-family: var(--font-mono);
}

.neo4j-sans {
  font-family: var(--font-sans);
}

/* Add any other app-wide styles here */ 