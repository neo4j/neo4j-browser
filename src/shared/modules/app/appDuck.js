/*
 * Copyright (c) 2002-2020 "Neo4j,"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

// Action type constants
export const NAME = 'app'
export const APP_START = `${NAME}/APP_START`
export const USER_CLEAR = `${NAME}/USER_CLEAR`

export const URL_ARGUMENTS_CHANGE = `${NAME}/URL_ARGUMENTS_CHANGE`

// State constants
export const DESKTOP = 'DESKTOP'
export const WEB = 'WEB'
export const CLOUD = 'CLOUD'

const SECURE_SCHEMES = ['neo4j+s', 'bolt+s']
const INSECURE_SCHEMES = ['neo4j', 'bolt']

// Selectors
export const getHostedUrl = state => (state[NAME] || {}).hostedUrl || null
export const getEnv = state => (state[NAME] || {}).env || WEB
export const hasDiscoveryEndpoint = state =>
  [WEB, CLOUD].includes(getEnv(state))
export const inWebEnv = state => getEnv(state) === WEB
export const inWebBrowser = state => [WEB, CLOUD].includes(getEnv(state))
export const getAllowedBoltSchemes = (state, encryptionFlag) => {
  const isHosted = inWebBrowser(state)
  const hostedUrl = getHostedUrl(state)
  return !isHosted
    ? encryptionFlag
      ? SECURE_SCHEMES
      : [...SECURE_SCHEMES, ...INSECURE_SCHEMES]
    : (hostedUrl || '').startsWith('https')
    ? SECURE_SCHEMES
    : INSECURE_SCHEMES
}

// Reducer
export default function reducer(state = { hostedUrl: null }, action) {
  switch (action.type) {
    case APP_START:
      return { ...state, hostedUrl: action.url, env: action.env }
    default:
      return state
  }
}
