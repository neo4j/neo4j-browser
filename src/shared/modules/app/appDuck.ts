/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { GlobalState } from 'shared/globalState'

// Action type constants
export const NAME = 'app'
export const APP_START = 'app/APP_START'
export const UPDATE_BUILD_INFO = 'app/UPDATE_BUILD_INFO'
export const USER_CLEAR = 'app/USER_CLEAR'
export type AppStartAction = { type: typeof APP_START }
export type UserClearAction = { type: typeof USER_CLEAR }

export const URL_ARGUMENTS_CHANGE = 'app/URL_ARGUMENTS_CHANGE'

// State constants
export const DESKTOP = 'DESKTOP'
export const WEB = 'WEB'
export const CLOUD = 'CLOUD'
export type Environment = typeof DESKTOP | typeof WEB | typeof CLOUD

export const SECURE_SCHEMES = ['neo4j+s', 'bolt+s']
export const INSECURE_SCHEMES = ['neo4j', 'bolt']
export const CLOUD_SCHEMES = ['neo4j+s']

// Selectors
export const getHostedUrl = (state: GlobalState): string | null =>
  (state[NAME] || {}).hostedUrl || null
export const getEnv = (state: GlobalState): Environment =>
  (state[NAME] || {}).env || WEB
export const hasDiscoveryEndpoint = (state: GlobalState): boolean =>
  [WEB, CLOUD].includes(getEnv(state))
export const inWebEnv = (state: GlobalState): boolean => getEnv(state) === WEB
export const inCloudEnv = (state: GlobalState): boolean =>
  getEnv(state) === CLOUD
export const inWebBrowser = (state: GlobalState): boolean =>
  [WEB, CLOUD].includes(getEnv(state))
export const inDesktop = (state: GlobalState): boolean =>
  getEnv(state) === DESKTOP

export const getAllowedBoltSchemes = (
  state: GlobalState,
  encryptionFlag?: any
) => {
  const isHosted = inWebBrowser(state)
  const hostedUrl = getHostedUrl(state)
  return !isHosted
    ? encryptionFlag
      ? SECURE_SCHEMES
      : [...SECURE_SCHEMES, ...INSECURE_SCHEMES]
    : (hostedUrl || '').startsWith('https')
    ? SECURE_SCHEMES
    : INSECURE_SCHEMES
}
// currently only Desktop specific
export const isRelateAvailable = (state: GlobalState): boolean =>
  Boolean(
    state[NAME].relateUrl &&
      state[NAME].relateApiToken &&
      state[NAME].relateProjectId
  )
export const getProjectId = (state: GlobalState): string | undefined =>
  state[NAME].relateProjectId

export type AppState = {
  hostedUrl?: string | null
  env?: Environment
  relateUrl?: string
  relateApiToken?: string
  relateProjectId?: string
  neo4jDesktopGraphAppId?: string
}

// Reducer
export default function reducer(
  state: AppState = { hostedUrl: null },
  action: any
): AppState {
  switch (action.type) {
    case APP_START:
      return {
        ...state,
        hostedUrl: action.url,
        env: action.env,
        relateUrl: action.relateUrl,
        relateApiToken: action.relateApiToken,
        relateProjectId: action.relateProjectId,
        neo4jDesktopGraphAppId: action.neo4jDesktopGraphAppId
      }
    default:
      return state
  }
}
