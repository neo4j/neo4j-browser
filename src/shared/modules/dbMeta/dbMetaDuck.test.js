/*
 * Copyright (c) 2002-2019 "Neo4j,"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import neo4j from 'neo4j-driver'
import reducer, * as meta from './dbMetaDuck'
import { APP_START } from 'shared/modules/app/appDuck'

describe('hydrating state', () => {
  test('should merge inital state and state on load', () => {
    // Given
    const action = { type: APP_START }

    // When
    const hydratedState = reducer({ foo: 'bar' }, action)

    // Then
    expect(hydratedState).toMatchSnapshot()
  })
})
describe('updating metadata', () => {
  test('should update state when metadata is updated', () => {
    const returnedLabels = {
      result: { name: 'labels' },
      get: () => {
        return { data: ['label1', 'label2'] }
      }
    }
    const returnedRelationshipTypes = {
      result: { name: 'relationshipTypes' },
      get: () => {
        return { data: ['rel1', 'rel2'] }
      }
    }
    const returnedProperties = {
      result: { name: 'properties' },
      get: () => {
        return { data: ['prop1', 'prop2'] }
      }
    }
    const returnedFunctions = {
      result: { name: 'functions' },
      get: () => {
        return {
          data: [
            {
              name: 'ns.functionName',
              signature: 'functionSignature',
              description: 'functionDescription'
            }
          ]
        }
      }
    }
    const returnedProcedures = {
      result: { name: 'procedures' },
      get: () => {
        return {
          data: [
            {
              name: 'ns.procedureName',
              signature: 'procedureSignature',
              description: 'procedureDescription'
            }
          ]
        }
      }
    }
    const returnedNodes = {
      result: { name: 'nodes' },
      get: () => ({
        data: neo4j.int(5)
      })
    }
    const returnedRelationships = {
      result: { name: 'relationships' },
      get: () => ({
        data: neo4j.int(10)
      })
    }

    const action = {
      type: meta.UPDATE_META,
      meta: {
        records: [
          returnedLabels,
          returnedRelationshipTypes,
          returnedProperties,
          returnedFunctions,
          returnedProcedures,
          returnedNodes,
          returnedRelationships
        ]
      },
      context: 'mycontext'
    }

    const nextState = reducer(undefined, action)

    expect(nextState.labels).toEqual([
      { val: 'label1', context: 'mycontext' },
      { val: 'label2', context: 'mycontext' }
    ])
    expect(nextState.relationshipTypes).toEqual([
      { val: 'rel1', context: 'mycontext' },
      { val: 'rel2', context: 'mycontext' }
    ])
    expect(nextState.properties).toEqual([
      { val: 'prop1', context: 'mycontext' },
      { val: 'prop2', context: 'mycontext' }
    ])
    expect(nextState.functions).toEqual([
      {
        val: 'ns.functionName',
        context: 'mycontext',
        signature: 'functionSignature',
        description: 'functionDescription'
      }
    ])
    expect(nextState.procedures).toEqual([
      {
        val: 'ns.procedureName',
        context: 'mycontext',
        signature: 'procedureSignature',
        description: 'procedureDescription'
      }
    ])
    expect(nextState.nodes).toEqual(5)
    expect(nextState.relationships).toEqual(10)
  })

  test('should update state with empty metadata', () => {
    const returnNothing = () => ({ data: [] })
    const returnNull = () => ({ data: null })
    const action = {
      type: meta.UPDATE_META,
      meta: {
        records: [
          { result: { name: 'labels' }, get: returnNothing },
          { result: { name: 'relationshipTypes' }, get: returnNothing },
          { result: { name: 'properties' }, get: returnNothing },
          { result: { name: 'functions' }, get: returnNothing },
          { result: { name: 'procedures' }, get: returnNothing },
          { result: { name: 'nodes' }, get: returnNull },
          { result: { name: 'realtionships' }, get: returnNull }
        ]
      },
      context: 'mycontext'
    }

    const nextState = reducer(undefined, action)

    expect(nextState.labels).toEqual([])
    expect(nextState.relationshipTypes).toEqual([])
    expect(nextState.properties).toEqual([])
    expect(nextState.functions).toEqual([])
    expect(nextState.procedures).toEqual([])
    expect(nextState.nodes).toEqual(0)
    expect(nextState.relationships).toEqual(0)
  })
  test('can update server settings', () => {
    // Given
    const initState = {
      shouldKeep: true
    }
    const action = {
      type: meta.UPDATE_SETTINGS,
      settings: {
        'browser.test': 1
      }
    }

    // When
    const nextState = reducer(initState, action)

    // Then
    expect(nextState).toMatchSnapshot()
  })

  test('can update server info', () => {
    // Given
    const initState = {
      shouldKeep: true
    }
    const action = {
      type: meta.UPDATE_SERVER,
      version: '3.2.0-RC2',
      edition: 'enterprise',
      storeId: 'xxxx'
    }

    // When
    const nextState = reducer(initState, action)

    // Then
    expect(nextState).toMatchSnapshot()
  })

  test('can CLEAR to reset state', () => {
    // Given
    const initState = {
      shouldKeep: false,
      server: {
        edition: 'enterprise',
        storeId: 'xxxx',
        version: '3.2.0-RC2'
      }
    }
    const action = {
      type: meta.CLEAR
    }

    // When
    const nextState = reducer(initState, action)

    // Then
    expect(nextState).toMatchSnapshot()
  })

  test('can update meta values with UPDATE', () => {
    // Given
    const initState = {
      myKey: 'val',
      noKey: true
    }
    const action = { type: meta.UPDATE, myKey: 'yo', secondKey: true }

    // When
    const nextState = reducer(initState, action)

    // Then
    expect(nextState).toMatchSnapshot()
  })
})
