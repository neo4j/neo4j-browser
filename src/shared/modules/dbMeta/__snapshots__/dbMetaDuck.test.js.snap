// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`updating metadata can CLEAR to reset state 1`] = `
Object {
  "functions": Array [],
  "labels": Array [],
  "procedures": Array [],
  "properties": Array [],
  "relationshipTypes": Array [],
  "server": Object {
    "edition": null,
    "storeId": null,
    "version": null,
  },
  "settings": Object {
    "browser.allow_outgoing_connections": false,
  },
}
`;

exports[`updating metadata can update server info 1`] = `
Object {
  "functions": Array [],
  "hydrated": true,
  "labels": Array [],
  "procedures": Array [],
  "properties": Array [],
  "relationshipTypes": Array [],
  "server": Object {
    "edition": "enterprise",
    "storeId": "xxxx",
    "version": "3.2.0-RC2",
  },
  "settings": Object {
    "browser.allow_outgoing_connections": false,
  },
  "shouldKeep": true,
}
`;

exports[`updating metadata can update server settings 1`] = `
Object {
  "functions": Array [],
  "hydrated": true,
  "labels": Array [],
  "procedures": Array [],
  "properties": Array [],
  "relationshipTypes": Array [],
  "server": Object {
    "edition": null,
    "storeId": null,
    "version": null,
  },
  "settings": Object {
    "browser.test": 1,
  },
  "shouldKeep": true,
}
`;
